src/Types.elm

```diff
@@ -2,18 +2,73 @@ module Types exposing (..)
 
 type Tree a = Node a (List (Tree a))
 
-type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole | TokVar String
-             | TokConstInt Int | TokConstBool Bool | TokInvalid | TokBackSlash | TokArrow | TokEnd
+type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole |
+              TokVar String | TokHasType | TokConstInt Int | TokConstBool Bool |
+              TokInvalid | TokBackSlash | TokArrow | TokEnd | TokTypeName String
 
 type TokTSC = TTSCInt TokTSCInt | TTSCBool TokTSCBool
 type TokTSCInt = TokPlus | TokMinus | TokTimes
 type TokTSCBool = TokEq | TokAnd | TokOr
 
 type Const = CBool Bool | CInt Int
 type Term = CTerm Const | VTerm String | Plus Term Term | Minus Term Term | Times Term Term
-            | Eq Term Term | And Term Term | Or Term Term | Lam Term Term | App Term Term
+            | Eq Term Term | And Term Term | Or Term Term | Lam String Term | App Term Term
             | MissingInt | MissingBool | Missing | EmptyTree
 
+-- V(alue)Type is a type that a TreeAssembly term can evaluate to
+{-
+  TAny may be used when the type can be either int or bool (discuss)
+  TNone is used as a default value for types we haven't determined yet (similar to EmptyTree)
+-}
```

Comment:
I hesitate to add `TAny` and `TNone` to our type system.  If we want to represent a term having multiple possible types then I think we should have some sort of set containing possible types (or a type variable).  As for `TNone`, I think it makes more sense to use a `Maybe VType` wherever a term may not have a valid type -- this is what existing code does, I believe.

Reply:
You're right, I just got rid of `TAny` (wasn't really using it), started using Maybe rather than `TAny`, and refactored environment's functions so that `TNone` is not necessary.